src/main/java/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -552,97 +540,86 @@ private static MoveValidationResult validateBasic(final GameData data,
         moveTest = MoveValidator.getNonLand(units);
-      } else {
-        moveTest = units;
       }
+      final Map<Unit, Collection<Unit>> dependentsMap =
+          getDependents(CollectionUtils.getMatches(units, Matches.unitCanTransport()));
+      Set<Unit> dependents = dependentsMap.values().stream().flatMap(Collection::stream).collect(Collectors.toSet());
+      dependents.addAll(newDependents.values().stream().flatMap(Collection::stream).collect(Collectors.toSet()));
+      moveTest.removeAll(dependents);
+
+      // Can only move owned units except transported units or allied air on carriers
       for (final Unit unit : CollectionUtils.getMatches(moveTest, Matches.unitIsOwnedBy(player).negate())) {
-        // allow allied fighters to move with carriers
-        if (!(UnitAttachment.get(unit.getType()).getCarrierCost() > 0 && data.getRelationshipTracker().isAllied(player,
-            unit.getOwner()))) {
+        if (!(UnitAttachment.get(unit.getType()).getCarrierCost() > 0
+            && data.getRelationshipTracker().isAllied(player, unit.getOwner()))) {
           result.addDisallowedUnit("Can only move own troops", unit);
         }
       }
-      // Initialize available Mechanized Inf support
-      int mechanizedSupportAvailable = getMechanizedSupportAvail(units, player);
-      final Map<Unit, Collection<Unit>> dependencies =
-          getDependents(CollectionUtils.getMatches(units, Matches.unitCanTransport()));
-      // add those just added
-      // TODO: do not EVER user something from the UI in a validation method. Only the local computer (ie: client) has a
-      // copy of this UI
-      // data. The server has a different copy!!!!
-      // TODO: re-write the entire fucking Paratroopers code. It is garbage! We need a single all encompassing UI and
-      // engine for all the
-      // different types of transportation that exist.
-      if (!newDependents.isEmpty()) {
-        for (final Unit transport : dependencies.keySet()) {
-          if (dependencies.get(transport).isEmpty()) {
-            dependencies.put(transport, newDependents.get(transport));
-          }
-        }
-      }
-      // check units individually
+
+      // Check if units have enough movement accounting for land transports
+      // Land transports can either:
+      // 1. Transport units on a 1-to-1 basis (have no capacity set)
+      // 2. Transport like sea transports using capacity and cost
       final Predicate<Unit> hasEnoughMovementForRoute;
       try {
         data.acquireReadLock();
         hasEnoughMovementForRoute = Matches.unitHasEnoughMovementForRoute(route);
       } finally {
         data.releaseReadLock();
       }
+      int numLandTransportsWithoutCapacity = getNumLandTransportsWithoutCapacity(units, player);
+      IntegerMap<Unit> landTransportsWithCapacity = getLandTransportsWithCapacity(units, player);
+      moveTest = TransportUtils.sortByTransportCostDescending(moveTest);
```

Comment:
The re-use of `moveTest` I think is very questionable. From here on out that variable has a new meaning, a new value. The code referring to `moveTest` above this point has a different meaning and value compared to the code below.

I think the only benefit we get is that we don't introduce a new name, but I would even suggest that is a drawback as the variable probably could use different names for the different things we are testing.

I suggest we create a new variable at this point, let `moveTest` be final, and have more descriptive variables as we check different things. Is that agreeable?